src/libraries/System.Private.CoreLib/src/System/TypedReference.cs

```diff
@@ -12,53 +12,54 @@ namespace System
     [CLSCompliant(false)]
     public ref partial struct TypedReference
     {
-
         public static TypedReference MakeTypedReference(object target, FieldInfo[] flds)
         {
             ArgumentNullException.ThrowIfNull(target);
             ArgumentNullException.ThrowIfNull(flds);
 
             if (flds.Length == 0)
+            {
                 throw new ArgumentException(SR.Arg_ArrayZeroError, nameof(flds));
+            }
+
+            ref byte targetRef = ref Unsafe.NullRef<byte>();
 
-            IntPtr[] fields = new IntPtr[flds.Length];
             // For proper handling of Nullable<T> don't change GetType() to something like 'IsAssignableFrom'
             // Currently we can't make a TypedReference to fields of Nullable<T>, which is fine.
             RuntimeType targetType = (RuntimeType)target.GetType();
             for (int i = 0; i < flds.Length; i++)
             {
                 RuntimeFieldInfo field = flds[i] as RuntimeFieldInfo ?? throw new ArgumentException(SR.Argument_MustBeRuntimeFieldInfo);
                 if (field.IsStatic)
+                {
                     throw new ArgumentException(SR.Format(SR.Argument_TypedReferenceInvalidField, field.Name));
+                }
 
                 if (targetType != field.GetDeclaringTypeInternal() && !targetType.IsSubclassOf(field.GetDeclaringTypeInternal()))
+                {
                     throw new MissingMemberException(SR.MissingMemberTypeRef);
+                }
 
                 RuntimeType fieldType = (RuntimeType)field.FieldType;
-                if (fieldType.IsPrimitive)
-                    throw new ArgumentException(SR.Format(SR.Arg_TypeRefPrimitive, field.Name));
-
                 if (i < (flds.Length - 1) && !fieldType.IsValueType)
+                {
                     throw new MissingMemberException(SR.MissingMemberNestErr);
-
-                fields[i] = field.FieldHandle.Value;
+                }
+
+                if (i == 0)
+                {
+                    targetRef = ref RuntimeFieldHandle.GetFieldDataReference(target, field);
+                }
+                else
+                {
+                    targetRef = ref RuntimeFieldHandle.GetFieldDataReference(ref targetRef, field);
+                }
```

Comment:
Can we somehow make it harder to call the wrong GetFieldDataReference by mistake? Or assert that we're not in an EnC added field when we call the `ref byte` one?

Reply:
Unclear if the EnC state is projected into managed code. Is there something I can add?